yaml-tests/src/main/java/com/apple/foundationdb/relational/yamltests/server/SemanticVersion.java

```diff
@@ -0,0 +1,193 @@
+    }
+
+    private static String dotSeparated(String part) {
+        return part + "(." + part + ")*";
+    }
+
+    @Override
+    public String toString() {
+        String version = versionNumbers.stream().map(Object::toString).collect(Collectors.joining("."));
+        if (!prerelease.isEmpty()) {
+            version = version + "-" + prerelease;
+        }
+        if (!buildMetadata.isEmpty()) {
+            version = version + "+" + buildMetadata;
+        }
+        return version;
+    }
+
+    @Override
+    public boolean equals(final Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        final SemanticVersion that = (SemanticVersion)o;
+        return Objects.equals(versionNumbers, that.versionNumbers) && Objects.equals(prerelease, that.prerelease) && Objects.equals(buildMetadata, that.buildMetadata);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(versionNumbers, prerelease, buildMetadata);
+    }
+
+    @Override
+    public int compareTo(SemanticVersion o) {
+        // negative if this is less than o
+        final int versionComparison = compareVersionNumbers(o);
+        if (versionComparison != 0) {
+            return versionComparison;
+        }
+        return comparePrerelease(o);
+    }
+
+    private int compareVersionNumbers(SemanticVersion o) {
+        // semver only allows 3, we only use 4.
+        // gradle supports an arbitrary number, but it gets complicated if there is a pre-release or build info
+        if (versionNumbers.size() != o.versionNumbers.size()) {
+            throw new IllegalArgumentException("Versions only differ by part count" + this + " vs " + o);
+        }
+        int length = versionNumbers.size();
+        for (int i = 0; i < length; i++) {
+            if (versionNumbers.get(i) < o.versionNumbers.get(i)) {
+                return -1;
+            } else if (versionNumbers.get(i) > o.versionNumbers.get(i)) {
+                return 1;
+            }
+        }
+        return 0;
+    }
+
+    private int comparePrerelease(SemanticVersion o) {
+        if (!prerelease.isEmpty() && o.prerelease.isEmpty()) {
+            return -1;
+        } else if (prerelease.isEmpty() && !o.prerelease.isEmpty()) {
+            return 1;
+        }
+        int length = Math.min(prerelease.size(), o.prerelease.size());
+        for (int i = 0; i < length; i++) {
+            String thisPart = prerelease.get(i);
+            String otherPart = o.prerelease.get(i);
+            if (NUMERIC_PATTERN.matcher(thisPart).matches()) {
+                if (NUMERIC_PATTERN.matcher(otherPart).matches()) {
+                    int thisNumber = Integer.parseInt(thisPart);
+                    int otherNumber = Integer.parseInt(otherPart);
+                    if (thisNumber < otherNumber) {
```

Comment:
It may add extra cost of boxing and unboxing, but maybe `Objects.compare()` here may be clearer?

Reply:
```
int result = Objects.compare(thisNumber, otherNumber);
if (result != 0) {
    result result;
}
```